File: app/libs/programming/controllerGenerator.py
import os
from .TemplateReader import TemplateReader

class ControllerGenerator:

	def __init__(self , database):
		self.database = database

	def generate(self):
		code = ""
		code += self.generateCode() + "\n"
		return code

	def generateCode(self):

		# get encrypt vars
		encryptVar = self.fetchAttributesEncrypt(self.database.getAttibutes())

		# get id vars
		idVars = self.fetchAttributesId(self.database.getAttibutes())

		# get validate function
		validateFunction = self.validateObj(encryptVar , idVars)

		# read prototype file
		templateReader = TemplateReader("/controller/controller.pt")
		key = {
			"class"				:	self.database.getName(),
			"validate" 			:	validateFunction
		}

		return templateReader.generate(key)

	def fetchAttributesEncrypt(self, attributes):
		att_list = []
		for att in attributes:
			if att.getIsEncrypt():
				att_list.append(att)
		return att_list

	def fetchAttributesId(self, attributes):
		att_list = []
		for att in attributes:
			if att.getIsId():
				att_list.append(att)
				break
		return att_list

	def validateObj(self , vars , idVars):

		if len(vars) > 0 and len(vars) < 2 and len(idVars) > 0:

			# read prototype file
			templateReader = TemplateReader("/controller/validate.pt")
			key = {
				"database"				:	self.database.getName(),
				"encrypt_var" 			:	vars[0].getName(),
				"encrypt_var_capital" 	:	vars[0].getName().capitalize(),
				"id_var" 				: 	idVars[0].getName()
			}

			return templateReader.generate(key)

		else:
			return ""

File: app/libs/programming/ControllerGenerator.js
const Template = require('../engine/FileTemplate')

class ControllerGenerator {

    constructor(database){
        this.database = database
        this.name = database.name
        this.keys = {}
        this.path = __dirname
    }

    generate(){

        this.setupClass()
        this.setupValidate()

        let template = new Template(`${this.path}/prototypes/controller/controller.pt`)
        return template.generate(this.keys)
    }

    setupClass(){

        this.keys = { ...this.keys , class:this.database.name }

    }

    setupValidate(){

        let encryptAtt = this.database.attributes.filter((att) => {
            return att.isEncrypt
        })

        let idAtt = this.database.attributes.filter((att) => {
            return att.isId
        })

        if (encryptAtt.length < 1 || idAtt.length < 1) return

        let template = new Template(`${this.path}/prototypes/controller/validate.pt`)
        let validate = template.generate({
            id_var: idAtt[0].name,
            encrypt_var: encryptAtt[0].name,
            encrypt_var_capital: this.capitalize(encryptAtt[0].name),
            database:this.database.name
        })

        this.keys = { ...this.keys , validate }

    }

    capitalize(text) {
        return text.charAt(0).toUpperCase() + text.slice(1);
    }

}

module.exports = ControllerGenerator